### PR TITLE
Add half math functions

### PR DESCRIPTION
**Pull request by e10harvey**
Related to https://github.com/kokkos/kokkos-kernels/pull/1774.

### CODE REVIEW

**Review by dalg24 — CHANGES_REQUESTED** [review]
These need to be added to the <Kokkos_Half.hpp> header

File: core/src/Kokkos_Half.hpp
Comment (lucbv) [inline core/src/Kokkos_Half.hpp:26]: Looks like this can be removed

File: core/src/Kokkos_Half.hpp
Comment (lucbv) [inline core/src/Kokkos_Half.hpp:1269]: Looks like you need an extra space somewhere to align this properly?
If it looks good in the documentation then its fine.

**Review by lucbv — COMMENTED** [review]
It looks fine to me just a couple clean ups needed

File: core/src/Kokkos_Half.hpp
Comment (dalg24) [inline core/src/Kokkos_Half.hpp:1074]: Please add the missing numerics trait in a separate PR

File: core/src/Kokkos_Half.hpp
Comment (dalg24) [inline core/src/Kokkos_Half.hpp:1459]:
```suggestion
KOKKOS_INLINE_FUNCTION Kokkos::Experimental::half_t imag(Kokkos::Experimental::half_t) {             
```

File: core/src/Kokkos_Half.hpp
Comment (dalg24) [inline core/src/Kokkos_Half.hpp:1368]: If you are going to cast to float then defer to Kokkos::FUNC

File: core/src/Kokkos_Half.hpp
Comment (dalg24) [inline core/src/Kokkos_Half.hpp:1370]:
```suggestion
    return static_cast<Kokkos::Experimental::half_t>(             \
        Kokkos::FUNC(static_cast<float>(x)));                             \
```
(not formatted)

File: core/src/Kokkos_Half.hpp
Comment (crtrott) [inline core/src/Kokkos_Half.hpp:1368]: Not needed, just call Kokkos::FUNC, after all you anyway call the float version.

File: core/src/Kokkos_Half.hpp
Comment (dalg24) [inline core/src/Kokkos_Half.hpp:1372]: Please comment about the suffixed versions that are being added.
Is there prior practice of using that suffix?  Do we need/want them?

Comment (crtrott) [inline core/src/Kokkos_Half.hpp:1372]: We probably shouldn't provide these: there is no clear convention for this (for example NVIDIA has hexp not exp_fp16, and uses hexp for both half and bhalf. So lets leave this overload out.

**Review by crtrott — CHANGES_REQUESTED** [review]
I also think we should split the file, into non-public subfiles and then include them all in Kokkos_Half.hpp

**Review by cwpearson — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

**Comment by e10harvey** [comment]
@dalg24, @crtrott: Ready to merge?

**Comment by dalg24** [comment]
> @dalg24, @crtrott: Ready to merge?

How does it relate to #6055 ?
There is no proper description and we are back to one large PR that does things can/should be separated out.

**Comment by e10harvey** [comment]
> > 
> 
> How does it relate to #6055 ?

Good question. This is simply related to 6055 in that the traits and infinity-helper were requested to be put in a separate PR (which is now 6055).

> There is no proper description and we are back to one large PR that does things can/should be separated out.

I am not sure I follow. The original motivation for this PR was to add the half-precision math functions and during review it was requested that the traits and math fns be moved into separate headers. Are you saying you want 3 PRs:
1. Move the floating_point_wrapper to a private header file
2. Move half traits to a private header file
3. Add half math fns to a private header file

If so, that's no problem. I can create 3 PRs.

File: core/src/impl/Kokkos_Half_MathematicalFunctions.hpp
Comment (crtrott) [inline core/src/impl/Kokkos_Half_MathematicalFunctions.hpp:31]: Is there any reason not to do half_t and bhalf_t in one go if both are available? Are there any math functions which don't work in both cases?

```c++
#define KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF_TYPE(FUNC, TYPE)          \
  KOKKOS_INLINE_FUNCTION TYPE FUNC( TYPE x)                          \
    { return static_cast<TYPE>(  Kokkos::FUNC(static_cast<float>(x)));  }
    
#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
  #if defined(KOKKOS_BHALF_T_IS_FLOAT) && !KOKKOS_BHALF_T_IS_FLOAT
    #define KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(FUNC) \
            KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF_TYPE(FUNC, Kokkos::Experimental::half_t) \
            KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF_TYPE(FUNC, Kokkos::Experimental::bhalf_t)
  #else
    #define KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(FUNC) \
         KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF_TYPE(FUNC, Kokkos::Experimental::half_t)
  #endif
#endif
```

This way we won't overlook any instantiation.

File: core/src/impl/Kokkos_Half_MathematicalFunctions.hpp
Comment (crtrott) [inline core/src/impl/Kokkos_Half_MathematicalFunctions.hpp:91]: In this case abs isn't different from the other Unary is it? Abs is a special case on the underlying math functions due to the way how abs is part of the pure integer stuff and the cmath in the standard, but it is not here a special case or?

**Review by crtrott — CHANGES_REQUESTED** [review]
I think we can do the mathematical functions simpler or? Otherwise looks ok.

**Comment by e10harvey** [comment]
@crtrott, @dalg24: I think I addressed your feedback. The CI failure in SYCL is unrelated to these changes. Ready to merge?

**Comment by dalg24** [comment]
> @crtrott, @dalg24: I think I addressed your feedback. The CI failure in SYCL is unrelated to these changes. Ready to merge?

As far as I am concerned no.  My objections are there is no proper description (you need to edit) and this patch is doing 3 things that really should be split up in 3 PRs.  I am not blocking if Christian decides your last changes addressed his concerns and he wants to merge.

At the moment you have two stales approval (approvals but for a version of the patch that is significantly out of date) and one request changes.  There is not enough consensus to merge yet.

**Comment by e10harvey** [comment]
Closing. Please review https://github.com/kokkos/kokkos/pull/6055. I will open 2 more PRs -- one for moving floating point wrapper to a private header and another for the half math functions.

**Pull request closed without merging**
